File: swift/213_HouseRobberII.swift
class Solution {
    func rob(_ nums: [Int]) -> Int {
        var N = nums.count
        if N == 1 { return nums[0] }
        func robPath(_ start: Int, _ end: Int) -> Int {
            var previousSum = 0, currentSum = 0
            for i in start...end {
                var temp = currentSum
                currentSum = max(previousSum + nums[i], currentSum)
                previousSum = temp
            }
            return currentSum
        }
        return max(robPath(0, N-2), robPath(1, N-1))
    }
}

File: cpp/213_HouseRobberII.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int N = nums.size();
        if (N == 1) return nums[0];
        
        return max(robPath(nums, 0, N-2), robPath(nums, 1, N-1));
    }
    
    int robPath(vector<int>& nums, int start, int end) {
        int prev = 0, high = 0;
        for (int i = start; i <= end; ++i) {
            int temp = high;
            high = max(prev + nums[i], high);
            prev = temp;
        }
        return high;
    }
};
